src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -4649,7 +4652,8 @@ internal unsafe void SetSocketOption(SocketOptionLevel optionLevel, SocketOption
             SocketError errorCode = SocketError.Success;
             try
             {
-                errorCode = SocketPal.SetSockOpt(_handle, optionLevel, optionName, optionValue);
+                if (!IOControlKeepAlive.TrySet(this, optionLevel, optionName, optionValue))
+                    errorCode = SocketPal.SetSockOpt(_handle, optionLevel, optionName, optionValue);
```

Comment:
Ditto.

Reply:
My initial implementation was in the SocketPal, but there I only have the `handle` therefore I cannot call `socket.IOControl` and I would have to call `WindowsIoctl` replicating the code of `IOControl`:

```csharp
public int IOControl(int ioControlCode, byte[] optionInValue, byte[] optionOutValue)
{
    if (CleanedUp)
        throw new ObjectDisposedException(this.GetType().FullName);
    int realOptionLength = 0;
    SocketError errorCode = SocketPal.WindowsIoctl(_handle, ioControlCode, optionInValue, optionOutValue, out realOptionLength);
    if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"Interop.Winsock.WSAIoctl returns errorCode:{errorCode}");
    if (errorCode != SocketError.Success)
        UpdateStatusAfterSocketErrorAndThrowException(errorCode);
    return realOptionLength;
}
```

I will try in `IOControlKeepAlive.TrySet` to replace `IOControl` with the following lines (propagating errors):
```csharp
int realOptionLength = 0;
SocketError errorCode = SocketPal.WindowsIoctl(_handle, ioControlCode, optionInValue, optionOutValue, out realOptionLength);
```

Reply:
> and I would have to call WindowsIoctl replicating the code of IOControl

That's fine.  All the CleanedUp / NetEventSource stuff could be dropped, and you already have separate error handling, so the duplication is minimal / almost nonexistent.